HomeWorks/HW5/ExpressionTree/ExpressionParser.cs

```diff
@@ -0,0 +1,140 @@
+using System;
+using System.IO;
+using System.Text;
+
+namespace ExpressionTree;
+
+/// <summary>
+/// Предоставляет функциональность для разбора арифметического выражения из строки.
+/// </summary>
+public class ExpressionParser
+{
+    private readonly StringReader reader;
+
+    /// <summary>
+    /// Инициализирует новый экземпляр класса <see cref="ExpressionParser"/>.
+    /// </summary>
+    /// <param name="input">Строка, содержащая арифметическое выражение.</param>
+    public ExpressionParser(string input)
+    {
+        reader = new StringReader(input);
+    }
+
+    /// <summary>
+    /// Разбирает арифметическое выражение.
+    /// </summary>
+    /// <returns>Дерево разбора выражения.</returns>
+    /// <exception cref="ParseException">Выбрасывается при ошибках разбора.</exception>
+    public Expression ParseExpression()
+    {
+        SkipWhiteSpace();
+        int next = reader.Peek();
+        if (next == -1)
+        {
+            throw new ParseException("Неожиданный конец входных данных.");
+        }
+
+        char ch = (char)next;
+        if (ch == '(')
+        {
+            // Потребляем открывающую скобку
+            reader.Read();
+            SkipWhiteSpace();
+
+            int opInt = reader.Read();
+            if (opInt == -1)
+            {
+                throw new ParseException("Неожиданный конец входных данных при ожидании оператора.");
+            }
+            char op = (char)opInt;
+            if (op != '+' && op != '-' && op != '*' && op != '/')
+            {
+                throw new ParseException($"Некорректный оператор '{op}'.");
+            }
+
+            Expression left = ParseExpression();
+            Expression right = ParseExpression();
+
+            SkipWhiteSpace();
+            int closing = reader.Read();
+            if (closing != ')')
+            {
+                throw new ParseException("Ожидалась закрывающая скобка ')'.");
+            }
+            return new BinaryOperationExpression(op, left, right);
+        }
+        else
+        {
+            return ParseNumber();
+        }
```

Comment:
```suggestion
       return ParseNumber();
```
В ветке if и так return в конце ведь

Reply:
Можно ещё уменьшить уровень вложенности, если проверку инвертировать и поместить return ParseNumber(); выше, в ветку if. Сейчас приходится всю ветку if пролистать, чтобы увидеть в конце метода одну строку